crates/bevy_input/src/input.rs

```diff
@@ -29,6 +29,35 @@ use bevy_ecs::schedule::State;
 /// * Using [`Input::clear_just_pressed`] or [`Input::clear_just_released`] instead.
 /// * Calling [`Input::clear`] or [`Input::reset`] immediately after the state change.
 ///
+/// ## Performance
+///
+/// For all operations, the following conventions are used:
+/// - **n** is the number of stored inputs.
+/// - **m** is the number of input types.
+/// - **\***-suffix denotes an amortized cost.
+/// - **~**-suffix denotes an expected cost.
+///
+/// See Rust's [std::collections doc on performance](https://doc.rust-lang.org/std/collections/index.html#performance) for more details on the conventions used here.
+///
+/// | **[`Input`] operations**          | **Computational complexity**       |
+/// |-----------------------------------|------------------------------------|
+/// | [`Input::any_just_pressed`]       | *O*(m*n)                           |
+/// | [`Input::any_just_released`]      | *O*(m*n)                           |
+/// | [`Input::any_pressed`]            | *O*(m*n)                           |
+/// | [`Input::get_just_pressed`]       | *O*(n)                             |
+/// | [`Input::get_just_released`]      | *O*(n)                             |
+/// | [`Input::get_pressed`]            | *O*(n)                             |
+/// | [`Input::just_pressed`]           | *O*(1)~                            |
+/// | [`Input::just_released`]          | *O*(1)~                            |
+/// | [`Input::pressed`]                | *O*(1)~                            |
+/// | [`Input::press`]                  | *O*(1)~*                           |
+/// | [`Input::release`]                | *O*(1)~*                           |
+/// | [`Input::release_all`]            | *O*(n)~*                           |
+/// | [`Input::clear_just_pressed`]     | *O*(1)~                            |
+/// | [`Input::clear_just_released`]    | *O*(1)~                            |
+/// | [`Input::reset_all`]              | *O*(n)                             |
+/// | [`Input::clear`]                  | *O*(n)                             |
```

Comment:
Is this accurate? HashSet::clear is O(1), no?

Reply:
I'll double-check properly when I can, and I could be wrong, but on clear the hashtable goes through all elements to drop them one at a time.
Maybe this is only true for types with a custom drop?

Reply:
HashSets are based on HashMaps, and both use an implementation based on the hasbrown crate.
If you follow down the clear() to where it is actually doing things, you get here:
https://github.com/rust-lang/hashbrown/blob/f2e62124cd947b5e2309dd6a24c7e422932aae97/src/raw/mod.rs#L1091
Which doesn't do much except call drop_elements which is defined here:
https://github.com/rust-lang/hashbrown/blob/f2e62124cd947b5e2309dd6a24c7e422932aae97/src/raw/mod.rs#L2369
That last one I consider to be `O(n)` hence the information put into the table. But its only run for cases where the type needs to be dropped.

So I guess I need to specify both cases? Or should we say it is `O(n)` worse case?

Reply:
I would go for `O(n)` because it's standard to refer to the worst case scenario when talking about time complexity.

Reply:
Looks like [it's marked as O(1)~ in the std docs](https://doc.rust-lang.org/stable/std/collections/index.html#maps).

I think it's more useful to know it's O(1) in 99.999999% of cases, and most likely in this case since there's probably less than 20 things in each of those sets.